File: src/Graphics/Cogh/Key.hs
module Graphics.Cogh.Key
  ( maybeKey
  )where

import Graphics.Cogh.CommonFFI

maybeKey :: Window -> IO (Maybe Event)
maybeKey = maybeEvent getKey $ \ p -> do
  code <- keyCode p
  isPress <- keyIsPress p
  isRepeat <- keyIsRepeat p
  return $ Key (KeyCode code) (cBool isPress) (cBool isRepeat)


foreign import ccall unsafe "getKey" getKey
  :: Window -> IO (Ptr ())

foreign import ccall unsafe "keyIsPress" keyIsPress
  :: Ptr () -> IO CInt

foreign import ccall unsafe "keyIsRepeat" keyIsRepeat
  :: Ptr () -> IO CInt

foreign import ccall unsafe "keyCode" keyCode
  :: Ptr () -> IO CUInt

File: csrc/key.c
#include "window-private.h"

#include "key.h"

Key** getKeys(Window* w) { return w->keys; }
int keyIsPress(Key* k) { return k->isPress; }
int keyIsRepeat(Key* k) { return k->isRepeat; }
unsigned int keyCode(Key* k) { return k->code; }

unsigned int readCode(const char* name) {
    return SDL_GetKeyFromName(name);
}

const char* showCode(unsigned int code) {
    return SDL_GetKeyName(code);
}
